src/main/java/com/iexec/core/task/TaskService.java

```diff
@@ -353,17 +353,26 @@ private void uploadRequested2UploadRequestTimeout(Task task) {
     }
 
     private void resultUploading2Uploaded(Task task) {
-        boolean condition1 = task.getCurrentStatus().equals(RESULT_UPLOADING);
-        Optional<Replicate> optionalReplicate = replicatesService.getReplicateWithResultUploadedStatus(task.getChainTaskId());
-        boolean condition2 = optionalReplicate.isPresent();
+        boolean isTaskInResultUploading = task.getCurrentStatus().equals(RESULT_UPLOADING);
+        Optional<Replicate> oUploadedReplicate = replicatesService.getReplicateWithResultUploadedStatus(task.getChainTaskId());
+        boolean didReplicateUpload = oUploadedReplicate.isPresent();
 
-        if (condition1 && condition2) {
-            task.setResultLink(optionalReplicate.get().getResultLink());
-            task.setChainCallbackData(optionalReplicate.get().getChainCallbackData());
+        if (!isTaskInResultUploading) {
+            return;
+        }
+
+        if (didReplicateUpload) {
+            task.setResultLink(oUploadedReplicate.get().getResultLink());
+            task.setChainCallbackData(oUploadedReplicate.get().getChainCallbackData());
             updateTaskStatusAndSave(task, RESULT_UPLOADED);
             resultUploaded2Finalized2Completed(task);
-        } else if (replicatesService.getNbReplicatesWithCurrentStatus(task.getChainTaskId(), ReplicateStatus.RESULT_UPLOAD_REQUEST_FAILED) > 0 &&
-                replicatesService.getNbReplicatesWithCurrentStatus(task.getChainTaskId(), ReplicateStatus.RESULT_UPLOADING) == 0) {
+            return;
+        }
+
+        boolean noReplicateIsUploading = replicatesService.getNbReplicatesWithCurrentStatus(
```

Comment:
We will have a problem if the worker is gone (WORKER_LOST, or worst RECOVERING) after RESUL_UPLOAD (but we can still merge if you bored!), I think we have a method like `getLastButOne` (?)
What do you think?

Reply:
No I'm never bored of this :)
If the worker is lost, then that's a good reason to request another upload, but you're right for the recovery case. I changed it in a way that we check if the worker is uploading or recovering to upload, otherwise another worker should upload.
Good point thanks!